### DISCUSSION TITLE
Add test coverage and type annotation requirement to quality scale

### DISCUSSION THREAD

**Discussion by farmio** [home-assistant/architecture#714] [General]
Hi 👋!

It would be nice if codeowners would be encouraged to reach 100% test coverage and have `mypy --strict` pass for their integrations. 
Although some libraries used by integrations may not be type-annotated or don't export their type hints I think having the integration code typed still adds to overall code quality and may help to prevent one or the other error.

## Proposal A

Add a new [quality scale](https://developers.home-assistant.io/docs/integration_quality_scale_index/) (for example "Star ⭐️") with requirements
- Satisfying all Platinum level requirements.
- Reach 100% test coverage across all integration files.
- Have fully type-annotated code and is added to `core/.strict-typing`

## Proposal B

Extend requirements of [quality scale "Platinum"](https://developers.home-assistant.io/docs/integration_quality_scale_index/#platinum-) with one or both of the above.

## Current State

As of 18. January 2022 there have been 40 platinum integrations of which 17 have 100% test coverage and 20 have strict type checking enabled.
100% test coverage for `config_flow.py` is required for all ConfigFlow enabled integrations - which is a "Gold" requirement.

**Comment by thecode** [comment]
Linking the quality scales: https://developers.home-assistant.io/docs/integration_quality_scale_index
I think the difference between `gold` and platinum is not big so **Proposal A** can be added to the requirements of `platinum`.

We should also consider adding strict typing to `gold` as it is not hard to implement and add 100% tests coverage for `platinum`

**Comment by frenck** [comment]
I'd like to suggest we add diagnostics to that as well.

Additionally, we should also do something with entity descriptions and provide a configuration_url when possible as well (the latter is probably more a gold addition?).

Sorry, I don't mean to de-rail this suggestion, however, I think it would be useful to centrally collect/discuss possible changes in general

**Comment by marvin-w** [comment]
To sum up what others have written here the current list as far as I understood. I also believe that adding the `entity_category` would make sense as well - maybe to gold?

## Gold 🥇

- ...
- Devices have a `configuration_url` (if appropriate)
- Entities have an appropriate `entity_category`
- Integration is strict type checked


## Platinum 🏆

- ...
- Integration supports gathering diagnostics
- Integration has full test coverage
- Integration uses entity descriptions (if appropriate)

**Reply by Kane610** [reply]
Strict typing should be on platinum. The problem is that platinum has got really hard requirements now. Should there be more levels or should stuff perhaps be moved around to even out the requirements?

**Reply by thecode** [reply]
Currently for gold it is required to have above average (97%) tests coverage for all modules, this is almost as saying full test coverage. I don't think there is any meaning to add the "Integration has full test coverage" to Platinum and I don't think requiring 100% is realistic.

Entity descriptions is not a hard requirement IMO and can be on gold. 

Strict type checking should be on Platinum.

**Reply by farmio** [reply]
I did not know gold has 97% coverage requirement. If that is the case it should be stated in the developer documentation site. 
"Above average" is a little vague. 

<hr>

The description for silver is 
> This integration is able to cope when things go wrong.

I think having diagnostics support would fit even there.

**Comment by thecode** [comment]
@farmio it is stated https://developers.home-assistant.io/docs/integration_quality_scale_index/#gold-
I added this change few days ago since it is the 2nd time I missed it. The average can't be fixed since it is depended on the current project coverage (as of now 97.09%).

**Comment by frenck** [comment]
This proposal is superseded by this one: <https://github.com/home-assistant/architecture/discussions/1098>

Which goes into a wide effort to revisit our scale. This includes the above-mentioned already.
